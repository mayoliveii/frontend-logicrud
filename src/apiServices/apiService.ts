/* eslint-disable @typescript-eslint/no-explicit-any */
import axios from 'axios';
import env from '../../env';

axios.create({
  baseURL: env.VITE_API_URL,
});

export interface NewCategoryData {
  title: string;
  description: string;
}

export interface EditCategoryData {
  id: string;
  title: string;
  description: string;
}

export interface CategoryData {
  id: string;
  title: string;
  description: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface ListCategoriesParams {
  orderBy?: 'created_at_ASC' | 'created_at_DESC';
  startDate?: string;
  endDate?: string;
  search?: string;
}

const api = axios.create({
  baseURL: env.VITE_API_URL,
});

const apiService = {
  getAllCategories: async (params: ListCategoriesParams): Promise<CategoryData[]> => {
    try {
      const response = await api.get<CategoryData[]>('/list-categories', { params });

      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  createCategory: async (categoryData: NewCategoryData) => {
    try {
      const response = await api.post('/create-category', categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  editCategory: async (categoryData: EditCategoryData) => {
    try {
      const response = await api.post(`/edit-category`, categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  getCategoryById: async (categoryId: string, categoryData: CategoryData) => {
    try {
      const response = await api.post(`/get-category-by-id?id=${categoryId}`, categoryData);
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  deleteCategory: async (categoryId: string) => {
    try {
      const response = await api.delete(`/delete-category`, { data: { id: categoryId } });
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
  searchCategorys: async (searchTerm: string, params: ListCategoriesParams) => {
    try {
      const response = await api.get('/list-categories', { params: { ...params, search: searchTerm } });
      return response.data;
    } catch (error) {
      handleApiError(error);
      throw error;
    }
  },
};

const handleApiError = (error: any) => {
  if (error.response) {
    throw new Error(`Error from server: ${JSON.stringify(error.response.data)}`);
  }

  if (error.request) {
    throw new Error(`No response from server: ${error.request}`);
  }

  if (error.message) {
    throw new Error(`Error: ${error.message}`);
  }
};

export default apiService;
