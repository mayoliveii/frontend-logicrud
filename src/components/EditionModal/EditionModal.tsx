/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  Button,
  useDisclosure,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalCloseButton,
  ModalBody,
  ModalFooter,
  Box,
  FormControl,
  FormLabel,
  Input,
  InputGroup,
  InputLeftElement,
  Textarea,
  useColorModeValue,
  VStack
} from '@chakra-ui/react';
import { FaTruck } from 'react-icons/fa';
import { useEffect, useState } from 'react';
import apiService, { CategoryData } from '../../apiServices/apiService';

interface EditModalProps {
  isOpen: boolean;
  onClose: () => void;
  category: CategoryData;
  onCategoryUpdated: () => void;
  handleSubmit?(): void;
}

export function EditionModal({ category, onCategoryUpdated }: EditModalProps) {
  const { isOpen, onOpen, onClose } = useDisclosure();

  const [title, setTitle] = useState(category.title);
  const [description, setDescription] = useState(category.description);

  const handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => setTitle(event.target.value);
  const handleDescriptionChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => setDescription(event.target.value);

  const handleSubmit = async () => {
    try {
      const updatedCategory = await apiService.editCategory({ id: category.id, title, description });
      setTitle('');
      setDescription('');

      onClose();
      if (onCategoryUpdated) {
        onCategoryUpdated();
      }

      return updatedCategory;
    } catch (error: any) {
      throw new Error(`Error updating category: ${error.message}`);
    }
  };

  useEffect(() => {
    if (isOpen) {
      setTitle(category.title);
      setDescription(category.description);
    }
  }, [isOpen, category]);
  return (
    <>
      <Button
        size={'md'}
        px={6}
        mr={4}
        onClick={onOpen}
        color={'white'}
        bg={'#FF7A50'}
        _hover={{ bg: '#FF5823' }}
      >
        editar
      </Button>

      <Modal isOpen={isOpen} onClose={onClose}>
        <ModalOverlay />
        <ModalContent>
          <ModalHeader>Adicionar nova categoria de veículo</ModalHeader>
          <ModalCloseButton />
          <ModalBody>
            <Box
              bg={useColorModeValue('white', 'gray.700')}
              borderRadius="lg"
              p={8}
              color={useColorModeValue('gray.700', 'whiteAlpha.900')}
              shadow="base">
              <VStack spacing={5}>
                <FormControl isRequired>
                  <FormLabel>Nome</FormLabel>

                  <InputGroup>
                    <InputLeftElement>
                      <FaTruck />
                    </InputLeftElement>
                    <Input type="text" value={title} onChange={handleNameChange} />
                  </InputGroup>
                </FormControl>

                <FormControl isRequired>
                  <FormLabel>Descrição</FormLabel>

                  <Textarea
                    rows={6}
                    resize="none"
                    value={description}
                    onChange={handleDescriptionChange}
                  />
                </FormControl>
              </VStack>
            </Box>
          </ModalBody>
          <ModalFooter>
            <Button bg='#FF7A50' mr={3} color={'white'} _hover={{ bg: '#FF5823' }} onClick={handleSubmit}>
              Salvar
            </Button>
            <Button variant='ghost' onClick={onClose}>cancelar</Button>
          </ModalFooter>
        </ModalContent>
      </Modal>
    </>
  );
}