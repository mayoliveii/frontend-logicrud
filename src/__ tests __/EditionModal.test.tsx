import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { ChakraProvider } from '@chakra-ui/react';
import { EditionModal } from '../components/EditionModal/EditionModal';
import { CategoryData } from '../apiServices/apiService';

jest.mock('./../apiServices/apiService', () => ({
  editCategory: jest.fn(),
}));

const onClose = jest.fn();
const onCategoryUpdated = jest.fn();

const editionModalPropsMock: CategoryData = {
  id: '1',
  title: 'Test',
  description: 'Test',
  created_at: '2021-09-15T00:00:00.000Z',
  updated_at: '2021-09-15T00:00:00.000Z',
  deleted_at: null
};

describe('EditionModal component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <ChakraProvider>
        <EditionModal isOpen={true} onClose={onClose} onCategoryUpdated={onCategoryUpdated} category={editionModalPropsMock} />
      </ChakraProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render when isOpen is false', () => {
    const { queryByText } = render(
      <ChakraProvider>
        <EditionModal isOpen={false} onClose={onClose} onCategoryUpdated={onCategoryUpdated} category={editionModalPropsMock} />
      </ChakraProvider>
    );
    expect(queryByText('Salvar')).not.toBeInTheDocument();
  });
});