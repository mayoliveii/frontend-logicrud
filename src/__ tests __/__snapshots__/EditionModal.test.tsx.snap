// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditionModal component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1uhuy6t"
    type="button"
  >
    editar
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
